### PR TITLE
Updating serviceErrorHandler to use apiVersion specific codec

### LINKED ISSUE #6999: ServiceErrorHandler should emit apiversion aware status
With https://github.com/GoogleCloudPlatform/kubernetes/pull/6881, we are now returning JSON with Status object, for paths that are not handled by the server.
But the Status object is encoded in the latest version. It should be encoded in the version that was requested by the user (which can be computed from the path).

We need to update go-restful for that, so that restful.ServiceErrorHandler is passed the request as well.

### PR DESCRIPTION
**Pull request by nikhiljindal**
Fixes https://github.com/GoogleCloudPlatform/kubernetes/issues/6999

```
$ curl -X PUT -i http://localhost:8080/api/v1beta3/redirect/namespaces/default/services/kubernetes
HTTP/1.1 405 Method Not Allowed
Content-Type: application/json
Date: Thu, 23 Apr 2015 07:12:57 GMT
Content-Length: 234

{
  "kind": "Status",
  "apiVersion": "v1beta3",
  "metadata": {},
  "status": "Failure",
  "message": "the server does not allow this method on the requested resource",
  "reason": "MethodNotAllowed",
  "details": {},
  "code": 405
}
```

```
curl -X PUT -i http://localhost:8080/api/v1beta2/redirect/services/kubernetes
HTTP/1.1 405 Method Not Allowed
Content-Type: application/json
Date: Thu, 23 Apr 2015 07:04:49 GMT
Content-Length: 245

{
  "kind": "Status",
  "creationTimestamp": null,
  "apiVersion": "v1beta2",
  "status": "Failure",
  "message": "the server does not allow this method on the requested resource",
  "reason": "MethodNotAllowed",
  "details": {},
  "code": 405
}
```

```
$ curl -X PUT -i http://localhost:8080/api/dsadas
HTTP/1.1 404 Not Found
Content-Type: application/json
Date: Thu, 23 Apr 2015 07:04:34 GMT
Content-Length: 211

{
  "kind": "Status",
  "apiVersion": "v1beta3",
  "metadata": {},
  "status": "Failure",
  "message": "the server could not find the requested resource",
  "reason": "NotFound",
  "details": {},
  "code": 404
}
```

### CODE REVIEW

**Comment by nikhiljindal** [comment]
cc @bgrant0607 @derekwaynecarr

**Comment by bgrant0607** [comment]
Thanks, @nikhiljindal.

@yujuhong PTAL. If you have questions, feel free to punt it back to me.

**Comment by yujuhong** [comment]
Ack. Looking now.

File: pkg/apiserver/apiserver.go [hunk 1]
```diff
@@ -171,13 +171,26 @@ func InstallLogsSupport(mux Mux) {
...
-	errorJSON(apierrors.NewGenericServerResponse(serviceErr.Code, "", "", "", "", 0, false), latest.Codec, response.ResponseWriter)
+func serviceErrorHandler(requestResolver *APIRequestInfoResolver, apiVersions []string, serviceErr restful.ServiceError, request *restful.Request, response *restful.Response) {
+	requestInfo, err := requestResolver.GetAPIRequestInfo(request.Request)
+	codec := latest.Codec
+	if err == nil && requestInfo.APIVersion != "" {
+		// check if the api version is valid.
+		for _, version := range apiVersions {
+			if requestInfo.APIVersion == version {
+				// valid api version.
+				codec = runtime.CodecFor(api.Scheme, requestInfo.APIVersion)
```
Comment (yujuhong) [inline pkg/apiserver/apiserver.go]: just a nit: you can `break` here

**Comment by yujuhong** [comment]
LGTM overall with a nit.

File: pkg/apiserver/apiserver.go (hunk 1 above)
Comment (nikhiljindal) [inline pkg/apiserver/apiserver.go]: Done. Thanks!

**Pull request merged**